src/lib/SaferNames/Name.hs

```diff
@@ -316,168 +234,339 @@ data ListE (e::E) (n::S) = ListE { fromListE :: [e n] }
-                          => FreshExt o o' -> Env (i:=>:i') (Name o')
-                          -> NameTraversal e m i  o
-                          -> NameTraversal e m i' o'
-extendInjectNameTraversal ext renamer t =
-  extendNameTraversal (injectNameTraversal ext t) renamer
-
-idRenamer :: HasNamesB b => b n l -> Env (n:=>:l) (Name l)
-idRenamer b = envMapWithKey (\name _ -> injectNameR name) $ boundScope b
-
--- variant with the common extension built in
-withTraverseNamesB
-  :: (HasNamesB b, Monad m)
-  =>                              Scope  o  -> RenameTraversal m  i  o  -> b i i'
-  -> (forall o'. FreshExt o o' -> Scope  o' -> RenameTraversal m  i' o' -> b o o' -> m a)
-  -> m a
-withTraverseNamesB s t b cont = do
-  traverseNamesB s t b >>= \case
-    FreshBinder ext b' renamer -> do
-      let t' = extendInjectNameTraversal ext renamer t
-      let s' = extendScope s b'
-      cont ext s' t' b'
-
-lookupNameTraversal :: Monad m => NameTraversal e m i o -> Name i
-                    -> m (Either (e o) (Name o))
-lookupNameTraversal (NameTraversal f env) name =
-  case projectName (envAsScope env) name of
-      Left  name' -> Left <$> f name'
-      Right name' -> return $ Right $ envLookup env name'
+class BindsNameList (b::B) (s::E) | b -> s where
+  (@@>) :: b i i' -> [v s o] -> EnvFrag v i i' o
+
+instance BindsOneName b s => BindsNameList (Nest b) s where
+  (@@>) Empty [] = emptyEnvFrag
+  (@@>) (Nest b rest) (x:xs) = b@>x <.> rest@@>xs
+  (@@>) _ _ = error "length mismatch"
+
+idSubst :: Typeable e => Typeable s => Env (SubstVal s e) n n
+idSubst = newEnv Rename
+
+applySubst :: ScopeReader m => SubstE v e => Env v i o -> e i -> m o (e o)
+applySubst env x = do
+  scope <- askScope
+  return $ runSubstReaderM scope env $ substE x
+
+applyAbs :: Typeable s => Typeable val
+         => ScopeReader m
+         => BindsOneName b s => SubstE (SubstVal s val) e
+         => Abs b e n -> val n -> m n (e n)
+applyAbs (Abs b body) x = applySubst (idSubst <>> (b@>SubstVal x)) body
+
+applyNaryAbs :: Typeable s => Typeable val
+             => ScopeReader m
+             => BindsNameList b s
+             => SubstE (SubstVal s val) e
+             => SubstB (SubstVal s val) b
+             => Abs b e n -> [val n] -> m n (e n)
+applyNaryAbs (Abs bs body) xs = applySubst (idSubst <>> (bs @@> map SubstVal xs)) body
+
+-- === versions of monad constraints with scope params ===
+
+type MonadKind  =           * -> *
+type MonadKind1 =      S -> * -> *
+type MonadKind2 = S -> S -> * -> *
+
+type Monad1 (m :: MonadKind1) = forall (n::S)        . Monad (m n  )
+type Monad2 (m :: MonadKind2) = forall (n::S) (l::S) . Monad (m n l)
+
+type MonadErr1 (m :: MonadKind1) = forall (n::S)        . MonadErr (m n  )
+type MonadErr2 (m :: MonadKind2) = forall (n::S) (l::S) . MonadErr (m n l)
+
+type MonadFail1 (m :: MonadKind1) = forall (n::S)        . MonadFail (m n  )
+type MonadFail2 (m :: MonadKind2) = forall (n::S) (l::S) . MonadFail (m n l)
```

Comment:
Can we unify `MonadErr` and `MonadFail`?

Reply:
I'm not sure what you mean. `MonadFail` is a standard haskell thing we need to implement if we want failable patterns in do-notation. And `MonadErr` carries our own custom error types.